File: crates/hyper_rhi/src/graphics_pipeline.rs
//
// Copyright (c) 2024, SkillerRaptor
//
// SPDX-License-Identifier: MIT
//

use std::{fmt::Debug, sync::Arc};

use downcast_rs::Downcast;

use crate::{pipeline_layout::PipelineLayout, shader_module::ShaderModule};

#[derive(Clone, Debug)]
pub struct GraphicsPipelineDescriptor<'a> {
    pub label: Option<&'a str>,
    pub layout: &'a Arc<dyn PipelineLayout>,
    pub vertex_shader: &'a Arc<dyn ShaderModule>,
    pub fragment_shader: &'a Arc<dyn ShaderModule>,
}

pub trait GraphicsPipeline: Debug + Downcast {}

downcast_rs::impl_downcast!(GraphicsPipeline);

File: hyper_rhi/src/hyper_rhi/graphics_pipeline.cpp
/*
 * Copyright (c) 2024, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "hyper_rhi/graphics_pipeline.hpp"

namespace hyper_rhi
{
    GraphicsPipeline::GraphicsPipeline(const GraphicsPipelineDescriptor &descriptor)
        : Resource(descriptor.label)
        , m_layout(descriptor.layout)
        , m_vertex_shader(descriptor.vertex_shader)
        , m_fragment_shader(descriptor.fragment_shader)
        , m_primitive_state(descriptor.primitive_state)
        , m_depth_stencil_state(descriptor.depth_stencil_state)
    {
    }

    PipelineLayout &GraphicsPipeline::layout() const
    {
        return *m_layout;
    }

    ShaderModule &GraphicsPipeline::vertex_shader() const
    {
        return *m_vertex_shader;
    }

    ShaderModule &GraphicsPipeline::fragment_shader() const
    {
        return *m_fragment_shader;
    }

    PrimitiveState GraphicsPipeline::primitive_state() const
    {
        return m_primitive_state;
    }

    DepthStencilState GraphicsPipeline::depth_stencil_state() const
    {
        return m_depth_stencil_state;
    }
} // namespace hyper_rhi